torch/_higher_order_ops/scan.py

```diff
@@ -0,0 +1,349 @@
+    assert len(args) == 2 * num_leaves
+    lhs = pytree.tree_unflatten(args[:num_leaves], spec)
+    rhs = pytree.tree_unflatten(args[num_leaves:], spec)
+    combined = combine_fn(lhs, rhs)
+    combined_leaves = pytree.tree_leaves(combined)
+    assert num_leaves == len(combined_leaves)
+    return combined_leaves
+
+
+def safe_map(f, *args):
+    args = list(map(list, args))
+    n = len(args[0])
+    for arg in args[1:]:
+        if len(arg) != n:
+            raise ValueError("length mismatch: {list(map(len, args))}")
+
+    def nf(a):
+        return f(*a)
+
+    return list(map(nf, zip(*args)))
+
+
+def scan(
+    combine_fn: Callable[[pytree.PyTree, pytree.PyTree], pytree.PyTree],
+    input: pytree.PyTree,
+    dim: int,
+    reverse: bool = False,
+    init: pytree.PyTree = [],  # noqa: B006
+) -> torch.Tensor:
+    r"""
+    Performs an inclusive scan with an associative pointwise combine function.
+
+    .. warning::
+        `torch.scan` is a prototype feature in PyTorch. It currently
+        does not support autograd and you may run into miscompiles.
+        Read more about feature classification at:
+        https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype
+
+    This operator requires runtime code generation and so requires support for
+    ``torch.compile``. Further, only CUDA device codegen is supported at the moment.
+
+    Args:
+        combine_fn (Callable): A binary callable with type ``(Tensor, Tensor) -> Tensor``,
+            or if input is a pytree ``(pytree, pytree) -> pytree``.
+            This function must be pure, pointwise, and satisfy the associative property.
+        input (torch.Tensor): The input tensor, or nested pytree of tensors.
+        dim (int): the dimension to scan over
+        reverse (bool): A boolean stating if the scan should be reversed with respect to the dimension.
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors.
+            The tensors are expected to have the same shape as the output tensors of ``combine_fn``
+
+    Example::
+
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        cumsum = scan(add, x, dim)
+
+    """
+    assert callable(combine_fn), "combine_fn must be a callable, but got {combine_fn}"
+    assert isinstance(dim, int), "dim must be an int, but got {type(dim)}"
+
+    # Dynamo is expecting a callable with "__code__" attribute.
+    # We cannot directly pass cond_op to it. So we wrap it in a dummy function.
+    def _scan_op_wrapper(*args, **kwargs):
+        return scan(*args, **kwargs)
+
+    if not torch._dynamo.is_compiling():
+        with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
+            return torch.compile(_scan_op_wrapper, backend="eager", fullgraph=True)(
+                combine_fn, input, dim=dim, reverse=reverse, init=init
+            )
+
+    leaves_init, spec_init = pytree.tree_flatten(init)
```

Comment:
1. we didn't have tests for pytree init values i think
2. torch.compile should be able to flatten the init and input, so here, we can expect them to be plain tuples.

Reply:
1. I added a testcase for pytree init values

2. I am not entirely sure what you mean? The `pytree.tree_flatten` is the same as we do for the `input`? I thought that this is the place where the pytree from the user is flattened into a plain tuple and the respective shape?

Reply:
I did a double check. Yeah, for scan, the pytrees are flattened in the wrapper code before entering dynamo. This is the opposite of cond or while_loop, where pytrees are flattened in dynamo. This is not ideal but I'm OK with the current status. Can you also add a `TODO: unify handing of pytrees for control flow ops`?

Reply:
This handling is the same as in `associative_scan`, if you prefer the way how cond or while_loop does it, we can switch it in a follow-up PR? 
For the moment, I added a [TODO](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/torch/_higher_order_ops/scan.py#L97) to `scan`

Reply:
This is OK.